File: studio/services/personalization_service.py
"""Personalized template generation from base templates + business profile."""

import html as html_lib
import logging
import os
import re
import time

from studio import draft
from studio.repositories import assets_repo, profiles_repo, projects_repo, templates_repo
from studio.services import composition_service, onboarding_service
from studio.services import component_rewrite_service
from studio.services.profile_facts import business_facts
from studio.services.profile_service import is_multi_place_value, split_multi_place_value
from studio.storage import s3

_log = logging.getLogger(__name__)

_PLACEHOLDER_RE = re.compile(r"\{\{[^}]+\}\}")
_EMAIL_TEXT_RE = re.compile(r"(?i)\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b")
_PHONE_TEXT_RE = re.compile(r"\+?\d[\d\s().-]{7,}\d")
_MAX_PACKS = 2
_STYLE_PREF_PACK_KEYWORDS = {
    "sleek-dark": ("sleek", "dark", "night", "moody"),
    "minimal-light": ("minimal", "clean", "simple", "light"),
    "bold-pop": ("bold", "pop", "vibrant", "neon", "colorful"),
    "frosted": ("glass", "frosted", "blur", "translucent"),
    "luxe-serif": ("luxe", "luxury", "premium", "elegant", "serif"),
    "high-contrast": ("high contrast", "contrast", "accessible"),
    "ocean-calm": ("ocean", "calm", "teal", "coastal", "aqua"),
}


class PersonalizationError(ValueError):
    """Raised when personalized HTML is invalid."""


def _brand(profile):
    name = (profile.get("brand") or {}).get("brandName")
    return html_lib.escape(name or "Your Brand", quote=True)


def _brand_slug(profile):
    brand = (profile.get("brand") or {}).get("brandName") or ""
    base = re.sub(r"[^a-z0-9]+", "-", str(brand).lower()).strip("-")
    return base or "yourbrand"


def _tagline(profile, genre):
    business = profile.get("business") or {}
    if business.get("description"):
        return html_lib.escape(business["description"], quote=True)
    services = business.get("services") or []
    btype = business.get("type") or "We"
    if services:
        text = f"{btype} — {', '.join(services[:3])}."
        return html_lib.escape(text, quote=True)
    default = draft.DEFAULT_TAGLINES.get(genre, draft.DEFAULT_TAGLINES["generic"])
    return html_lib.escape(default, quote=True)


def _location_display_text(profile):
    location = profile.get("location") or {}
    country = str(location.get("country") or "").strip()
    city = str(location.get("city") or "").strip()
    street = str(location.get("address") or "").strip()

    if country and is_multi_place_value(country):
        names = ", ".join(split_multi_place_value(country))
        return f"We work in many countries — {names}"

    if city and is_multi_place_value(city):
        names = ", ".join(split_multi_place_value(city))
        return f"Serving {names}"

    if street and city:
        return ", ".join([street, city])
    if city:
        return city
    if country:
        return country
    return "Your city"


def _apply_contact(html, profile):
    contact = profile.get("contact") or {}
    location = profile.get("location") or {}
    raw_email = str(contact.get("email") or "").strip()
    if not raw_email:
        raw_email = f"hello@{_brand_slug(profile)}.site"
    raw_phone = str(contact.get("phone") or "").strip() or "Add your phone number"

    email = html_lib.escape(raw_email, quote=True)
    phone = html_lib.escape(raw_phone, quote=True)
    phone_href = re.sub(r"\D", "", raw_phone)
    if phone_href and raw_phone.startswith("+"):
        phone_href = f"+{phone_href}"
    phone_href = html_lib.escape(phone_href, quote=True)
    address = html_lib.escape(_location_display_text(profile), quote=True)

    html = re.sub(r"(?i)\bAdd your number here\b", phone, html)
    html = re.sub(r"(?i)\bAdd your phone number\b", phone, html)
    html = re.sub(r"(?i)\bAdd your phone\b", phone, html)
    html = re.sub(r"(?i)hello@yourdomain\.com", email, html)
    html = re.sub(r"(?i)hello@[A-Za-z0-9.-]+\.example", email, html)
    html = re.sub(r"(?i)hello@example\.com", email, html)
    html = re.sub(r"Your street, your city", address, html)

    # Normalize any template-provided contact literals.
    html = _EMAIL_TEXT_RE.sub(email, html)
    html = _PHONE_TEXT_RE.sub(phone, html)

    html = re.sub(
        r'(?i)(href\s*=\s*["\'])mailto:[^"\']+(["\'])',
        rf"\1mailto:{email}\2",
        html,
    )
    if phone_href:
        html = re.sub(
            r'(?i)(href\s*=\s*["\'])tel:[^"\']*(["\'])',
            rf"\1tel:{phone_href}\2",
            html,
        )

    lower = html.lower()
    has_email = email.lower() in lower or f"mailto:{email.lower()}" in lower
    has_phone = phone.lower() in lower or (bool(phone_href) and f"tel:{phone_href.lower()}" in lower)
    if has_email and has_phone:
        return html

    phone_line = (
        f'<a href="tel:{phone_href}">{phone}</a>' if phone_href else f"{phone}"
    )
    section = (
        '<section id="contact" class="builder-contact" '
        'style="padding:48px 24px;border-top:1px solid #e5e7eb;">'
        '<div style="max-width:960px;margin:0 auto;">'
        '<h2 style="margin:0 0 12px;">Contact</h2>'
        f'<p style="margin:0 0 8px;">Email: <a href="mailto:{email}">{email}</a></p>'
        f"<p style=\"margin:0 0 8px;\">Phone: {phone_line}</p>"
        f'<p style="margin:0;">Location: {address}</p>'
        "</div></section>"
    )
    if re.search(r"</body>", html, flags=re.I):
        return re.sub(r"</body>", f"{section}</body>", html, count=1, flags=re.I)
    return f"{html}{section}"


def _apply_services(html, profile):
    services = (profile.get("business") or {}).get("services") or []
    for idx, service in enumerate(services[:3]):
        safe = html_lib.escape(service, quote=True)
        html = html.replace("Your main offering", safe, 1)
        html = html.replace("Your second strength", safe, 1)
        html = html.replace("The specialist request", safe, 1)
    audience = (profile.get("business") or {}).get("targetAudience")
    if audience:
        safe = html_lib.escape(audience, quote=True)
        html = html.replace("people nearby", safe, 1)
    return html


def _logo_url(assets):
    for asset in assets:
        if asset.get("assetType") == "logo" and asset.get("status") == "ready":
            key = asset.get("s3Key", "")
            if key:
                return s3.public_asset_url(key)
    return None


def _apply_logo(html, assets):
    logo = _logo_url(assets)
    if not logo:
        return html
    safe = html_lib.escape(logo, quote=True)
    if "<img" in html:
        return re.sub(
            r'(<img[^>]+src=")[^"]+(")',
            rf"\1{safe}\2",
            html,
            count=1,
        )
    return html.replace(
        '<a class="brand"',
        f'<img src="{safe}" alt="logo" style="height:32px;margin-right:8px;" /><a class="brand"',
        1,
    )


# Wording must not match draft._VISUAL_STYLE_HINT_RE, or refine() unlocks restyling.
_COPY_REWRITE_PROMPT = (
    "Rewrite the visible text of this page (headings, eyebrows, buttons, nav "
    "links, section copy, footer text) so every line accurately describes this "
    "business:\n{facts}\n"
    "Remove or replace copy about products this business does not offer. "
    "Keep the brand name and contact details exactly as they are. "
    "Change text only; keep every section and its structure."
)


def _rewrite_copy(html, profile):
    """LLM pass that grounds template copy in the business profile.

    Falls back to the string-substituted html when no provider is configured
    or the model returns an invalid document.
    """
    facts = business_facts(profile)
    if not facts:
        return html
    provider, model = onboarding_service._load_onboarding_provider()
    if provider is None or not model:
        return html
    try:
        rewritten = draft.refine(
            provider, model, html, _COPY_REWRITE_PROMPT.format(facts=facts)
        )
    except Exception:
        return html
    return rewritten or html


def personalize_html(raw_html, profile, assets, genre):
    html = raw_html.replace("{{BRAND}}", _brand(profile))
    html = html.replace("{{TAGLINE}}", _tagline(profile, genre))
    html = _apply_contact(html, profile)
    html = _apply_services(html, profile)
    html = _apply_logo(html, assets)
    html = draft.ensure_loadable_images(html, genre)
    if _PLACEHOLDER_RE.search(html):
        raise PersonalizationError("unresolved template placeholders remain")
    html = draft.normalize_cta_anchors(html)
    return draft.sanitize_html(html)


def _genre_hint(project, profile):
    prompt = project.get("initialPrompt") or ""
    business = profile.get("business") or {}
    parts = [
        prompt,
        business.get("type") or "",
        " ".join(business.get("services") or []),
        business.get("targetAudience") or "",
    ]
    return " ".join(p for p in parts if p).strip() or "generic website"


def _persist_key(project_id, template_id):
    return f"projects/{project_id}/templates/{template_id}.html"


def _preferred_pack_ids(profile):
    pref = str(((profile.get("design") or {}).get("stylePreference") or "")).lower().strip()
    if not pref:
        return []
    scored = []
    for idx, pack in enumerate(draft.STYLE_PACKS[1:]):
        pid = pack["id"]
        keywords = _STYLE_PREF_PACK_KEYWORDS.get(pid, ())
        score = sum(1 for kw in keywords if kw in pref)
        if score:
            scored.append((-score, idx, pid))
    scored.sort()
    return [pid for _, __, pid in scored]


def _selected_style_packs(profile):
    selected = []
    by_id = {pack["id"]: pack for pack in draft.STYLE_PACKS[1:]}
    for pid in _preferred_pack_ids(profile):
        pack = by_id.get(pid)
        if pack and pack not in selected:
            selected.append(pack)
        if len(selected) >= _MAX_PACKS:
            return selected
    for pack in draft.STYLE_PACKS[1:]:
        if pack not in selected:
            selected.append(pack)
        if len(selected) >= _MAX_PACKS:
            break
    return selected


def _composed_count():
    raw = os.environ.get("STUDIO_COMPOSED_VARIANTS", "2")
    try:
        n = int(raw)
    except ValueError:
        n = 2
    return max(0, min(n, 3))


def _composed_templates(project_id, profile, assets, genre):
    """Composed variants as template dicts. Any failure returns what succeeded."""
    out = []
    section_ms = 0.0
    try:
        composed = composition_service.compose_project_variants(
            project_id, business_facts(profile), genre, _composed_count()
        )
        for idx, comp in enumerate(composed):
            try:
                html = personalize_html(comp["html"], profile, assets, genre)
                if idx == 0:
                    rw_start = time.perf_counter()
                    if os.environ.get("STUDIO_COMPONENT_REWRITE", "1").strip().lower() in (
                        "0",
                        "false",
                        "no",
                    ):
                        html = _rewrite_copy(html, profile)
                    else:
                        html = component_rewrite_service.rewrite_components_parallel(
                            html, profile
                        )
                    section_ms = (time.perf_counter() - rw_start) * 1000
            except PersonalizationError:
                _log.warning("composed variant %s failed personalization", idx)
                continue
            template_id = f"composed-{idx + 1}"
            out.append(
                {
                    "templateId": template_id,
                    "label": f"Composed {idx + 1} · {genre.title()}",
                    "style": genre,
                    "sourceTemplateRef": ",".join(comp["componentIds"]),
                    "s3HtmlKey": _persist_key(project_id, template_id),
                    "htmlContent": html,
                }
            )
    except Exception:
        _log.exception("composed variants skipped for project %s", project_id)
    return out, section_ms


def generate_for_project(project_id, *, force=False):
    started = time.perf_counter()
    project = projects_repo.get(project_id)
    if not project:
        raise ValueError("project not found")
    existing = templates_repo.list_for_project(project_id)
    if existing and not force:
        return existing
    profile = profiles_repo.get(project_id)
    assets = assets_repo.list_for_project(project_id)
    genre = draft.pick_template(_genre_hint(project, profile))
    templates = []
    compose_ms = 0.0
    section_ms = 0.0

    compose_started = time.perf_counter()
    composed, section_ms = _composed_templates(project_id, profile, assets, genre)
    compose_ms = (time.perf_counter() - compose_started) * 1000 - section_ms
    if composed:
        for i, t in enumerate(composed):
            t["sourceKind"] = "composed"
            t["galleryIndex"] = i
            templates.append(t)
        base_html = composed[0]["htmlContent"]
        for j, pack in enumerate(_selected_style_packs(profile)):
            packed = draft._apply_pack(base_html, pack)
            if _PLACEHOLDER_RE.search(packed):
                raise PersonalizationError("unresolved placeholders in style pack")
            template_id = f"{genre}-{pack['id']}"
            templates.append(
                {
                    "templateId": template_id,
                    "label": f"{pack['label']} · {genre.title()}",
                    "style": genre,
                    "sourceTemplateRef": composed[0].get("sourceTemplateRef", ""),
                    "s3HtmlKey": _persist_key(project_id, template_id),
                    "htmlContent": packed,
                    "sourceKind": "pack",
                    "galleryIndex": len(composed) + j,
                }
            )
    else:
        design_files = draft.template_files(genre)
        if design_files:
            fname = design_files[0]
            stem = fname[: -len(".html")]
            with open(os.path.join(draft.TEMPLATES_DIR, fname), encoding="utf-8") as f:
                raw = f.read()
            html = personalize_html(raw, profile, assets, genre)
            templates.append(
                {
                    "templateId": stem,
                    "label": draft._design_label(raw, stem, genre),
                    "style": genre,
                    "sourceTemplateRef": fname,
                    "s3HtmlKey": _persist_key(project_id, stem),
                    "htmlContent": html,
                    "sourceKind": "fallback",
                    "galleryIndex": 0,
                }
            )

    templates.sort(key=lambda t: (t.get("galleryIndex", 999), t.get("templateId", "")))

    saved = templates_repo.replace_for_project(project_id, templates)
    projects_repo.update_fields(
        project_id,
        {"status": "ready", "templateCount": len(saved)},
    )
    _log.info(
        "generate_timing project=%s compose_ms=%.1f section_batch_ms=%.1f templates=%d total_ms=%.1f",
        project_id,
        compose_ms,
        section_ms,
        len(saved),
        (time.perf_counter() - started) * 1000,
    )
    return saved

File: src/server/services/personalizationService.ts
import fs from "node:fs";
import path from "node:path";
import {
  applyPack,
  DEFAULT_TAGLINES,
  designLabel,
  ensureLoadableImages,
  normalizeCtaAnchors,
  pickTemplate,
  sanitizeHtml,
  STYLE_PACKS,
  templateFiles,
  TEMPLATES_DIR,
  type StylePack,
} from "../draft";
import { businessFacts } from "../profileFacts";
import * as assetsRepo from "../repos/assetsRepo";
import * as profilesRepo from "../repos/profilesRepo";
import * as projectsRepo from "../repos/projectsRepo";
import * as templatesRepo from "../repos/templatesRepo";
import { publicAssetUrl } from "../storage/s3";
import { isMultiPlaceValue, splitMultiPlaceValue } from "./profileService";
import * as compositionService from "./compositionService";

export class PersonalizationError extends Error {}

const PLACEHOLDER_RE = /\{\{[^}]+\}\}/;
const EMAIL_TEXT_RE = /\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/gi;
const PHONE_TEXT_RE = /\+?\d[\d\s().-]{7,}\d/g;
const TAG_SPLIT_RE =
  /(<script\b[\s\S]*?<\/script\s*>|<style\b[\s\S]*?<\/style\s*>|<[^>]*>)/gi;
const MAX_PACKS = 2;

const STYLE_PREF_PACK_KEYWORDS: Record<string, string[]> = {
  "sleek-dark": ["sleek", "dark", "night", "moody"],
  "minimal-light": ["minimal", "clean", "simple", "light"],
  "bold-pop": ["bold", "pop", "vibrant", "neon", "colorful"],
  frosted: ["glass", "frosted", "blur", "translucent"],
  "luxe-serif": ["luxe", "luxury", "premium", "elegant", "serif"],
  "high-contrast": ["high contrast", "contrast", "accessible"],
  "ocean-calm": ["ocean", "calm", "teal", "coastal", "aqua"],
};

function escapeHtml(value: string): string {
  return value
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

function brand(profile: Record<string, unknown>): string {
  const name = (profile.brand as Record<string, unknown> | undefined)?.brandName;
  return escapeHtml(String(name || "Your Brand"));
}

function brandSlug(profile: Record<string, unknown>): string {
  const brandName = String((profile.brand as Record<string, unknown> | undefined)?.brandName ?? "");
  const base = brandName.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
  return base || "yourbrand";
}

function tagline(profile: Record<string, unknown>, genre: string): string {
  const business = (profile.business as Record<string, unknown> | undefined) ?? {};
  if (business.description) return escapeHtml(String(business.description));
  const services = Array.isArray(business.services) ? business.services : [];
  const btype = String(business.type ?? "We");
  if (services.length) {
    return escapeHtml(`${btype} — ${services.slice(0, 3).join(", ")}.`);
  }
  return escapeHtml(DEFAULT_TAGLINES[genre] ?? DEFAULT_TAGLINES.generic!);
}

function locationDisplayText(profile: Record<string, unknown>): string {
  const location = (profile.location as Record<string, unknown> | undefined) ?? {};
  const country = String(location.country ?? "").trim();
  const city = String(location.city ?? "").trim();
  const street = String(location.address ?? "").trim();
  if (country && isMultiPlaceValue(country)) {
    return `We work in many countries — ${splitMultiPlaceValue(country).join(", ")}`;
  }
  if (city && isMultiPlaceValue(city)) {
    return `Serving ${splitMultiPlaceValue(city).join(", ")}`;
  }
  if (street && city) return [street, city].join(", ");
  if (city) return city;
  if (country) return country;
  return "Your city";
}

function subVisibleText(html: string, pattern: RegExp, repl: string): string {
  const parts = html.split(TAG_SPLIT_RE);
  for (let i = 0; i < parts.length; i += 2) {
    parts[i] = parts[i]!.replace(pattern, () => repl);
  }
  return parts.join("");
}

function applyContact(html: string, profile: Record<string, unknown>): string {
  const contact = (profile.contact as Record<string, unknown> | undefined) ?? {};
  const rawEmail = String(contact.email ?? "").trim() || `hello@${brandSlug(profile)}.site`;
  const rawPhone = String(contact.phone ?? "").trim() || "Add your phone number";
  const email = escapeHtml(rawEmail);
  const phone = escapeHtml(rawPhone);
  let phoneHref = rawPhone.replace(/\D/g, "");
  if (phoneHref && rawPhone.startsWith("+")) phoneHref = `+${phoneHref}`;
  phoneHref = escapeHtml(phoneHref);
  const address = escapeHtml(locationDisplayText(profile));

  let out = html
    .replace(/\bAdd your number here\b/gi, phone)
    .replace(/\bAdd your phone number\b/gi, phone)
    .replace(/\bAdd your phone\b/gi, phone)
    .replace(/hello@yourdomain\.com/gi, email)
    .replace(/hello@[A-Za-z0-9.-]+\.example/gi, email)
    .replace(/hello@example\.com/gi, email)
    .replace(/Your street, your city/g, address);
  out = subVisibleText(out, EMAIL_TEXT_RE, email);
  out = subVisibleText(out, PHONE_TEXT_RE, phone);
  out = out.replace(/(href\s*=\s*["'])mailto:[^"']+(["'])/gi, `$1mailto:${email}$2`);
  if (phoneHref) out = out.replace(/(href\s*=\s*["'])tel:[^"']*(["'])/gi, `$1tel:${phoneHref}$2`);

  const lower = out.toLowerCase();
  const hasEmail = lower.includes(email.toLowerCase()) || lower.includes(`mailto:${email.toLowerCase()}`);
  const hasPhone =
    lower.includes(phone.toLowerCase()) || (Boolean(phoneHref) && lower.includes(`tel:${phoneHref.toLowerCase()}`));
  if (hasEmail && hasPhone) return out;

  const phoneLine = phoneHref
    ? `<a href="tel:${phoneHref}" style="color:var(--accent,#0d6efd);">${phone}</a>`
    : phone;
  const section =
    '<section id="contact" data-section="contact" class="builder-contact" ' +
    'style="padding:64px 24px;background:var(--bg,#fff);color:var(--ink,#111);' +
    'border-top:1px solid var(--line,#e5e7eb);font:inherit;">' +
    '<div class="container" style="max-width:1140px;margin:0 auto;">' +
    '<h2 style="margin:0 0 16px;color:inherit;font-size:1.75rem;">Contact</h2>' +
    `<p style="margin:0 0 8px;color:inherit;">Email: <a href="mailto:${email}" style="color:var(--accent,#0d6efd);">${email}</a></p>` +
    `<p style="margin:0 0 8px;color:inherit;">Phone: ${phoneLine}</p>` +
    `<p style="margin:0;color:inherit;">Location: ${address}</p></div></section>`;
  if (/<footer\b/i.test(out)) return out.replace(/<footer\b/i, `${section}<footer`);
  if (/<\/body>/i.test(out)) return out.replace(/<\/body>/i, `${section}</body>`);
  return `${out}${section}`;
}

function applyServices(html: string, profile: Record<string, unknown>): string {
  const services = ((profile.business as Record<string, unknown> | undefined)?.services as string[]) ?? [];
  let out = html;
  for (const service of services.slice(0, 3)) {
    const safe = escapeHtml(service);
    out = out.replace("Your main offering", safe);
    out = out.replace("Your second strength", safe);
    out = out.replace("The specialist request", safe);
  }
  const audience = (profile.business as Record<string, unknown> | undefined)?.targetAudience;
  if (audience) out = out.replace("people nearby", escapeHtml(String(audience)));
  return out;
}

function logoUrl(assets: Record<string, unknown>[]): string | null {
  for (const asset of assets) {
    if (asset.assetType === "logo" && asset.status === "ready" && asset.s3Key) {
      return publicAssetUrl(String(asset.s3Key));
    }
  }
  return null;
}

function applyLogo(html: string, assets: Record<string, unknown>[]): string {
  const logo = logoUrl(assets);
  if (!logo) return html;
  const safe = escapeHtml(logo);
  if (html.includes("<img")) {
    return html.replace(/(<img[^>]+src=")[^"]+(")/, `$1${safe}$2`);
  }
  return html.replace(
    '<a class="brand"',
    `<img src="${safe}" alt="logo" style="height:32px;margin-right:8px;" /><a class="brand"`,
  );
}

export function personalizeHtml(
  rawHtml: string,
  profile: Record<string, unknown>,
  assets: Record<string, unknown>[],
  genre: string,
): string {
  let html = rawHtml.replace(/\{\{BRAND\}\}/g, brand(profile));
  html = html.replace(/\{\{TAGLINE\}\}/g, tagline(profile, genre));
  html = applyContact(html, profile);
  html = applyServices(html, profile);
  html = applyLogo(html, assets);
  html = ensureLoadableImages(html, genre);
  if (PLACEHOLDER_RE.test(html)) throw new PersonalizationError("unresolved template placeholders remain");
  html = normalizeCtaAnchors(html);
  return sanitizeHtml(html);
}

function genreHint(project: Record<string, unknown>, profile: Record<string, unknown>): string {
  const prompt = String(project.initialPrompt ?? "");
  const business = (profile.business as Record<string, unknown> | undefined) ?? {};
  const parts = [
    prompt,
    String(business.type ?? ""),
    Array.isArray(business.services) ? business.services.join(" ") : "",
    String(business.targetAudience ?? ""),
  ].filter(Boolean);
  return parts.join(" ").trim() || "generic website";
}

function persistKey(projectId: string, templateId: string): string {
  return `projects/${projectId}/templates/${templateId}.html`;
}

function preferredPackIds(profile: Record<string, unknown>): string[] {
  const pref = String(((profile.design as Record<string, unknown> | undefined)?.stylePreference ?? "")).toLowerCase();
  if (!pref) return [];
  const scored: Array<[number, number, string]> = [];
  for (let idx = 0; idx < STYLE_PACKS.length - 1; idx++) {
    const pack = STYLE_PACKS[idx + 1]!;
    const keywords = STYLE_PREF_PACK_KEYWORDS[pack.id] ?? [];
    const score = keywords.filter((kw) => pref.includes(kw)).length;
    if (score) scored.push([-score, idx, pack.id]);
  }
  scored.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  return scored.map(([, , pid]) => pid);
}

function selectedStylePacks(profile: Record<string, unknown>): StylePack[] {
  const selected: StylePack[] = [];
  const byId = Object.fromEntries(STYLE_PACKS.slice(1).map((p) => [p.id, p]));
  for (const pid of preferredPackIds(profile)) {
    const pack = byId[pid];
    if (pack && !selected.includes(pack)) selected.push(pack);
    if (selected.length >= MAX_PACKS) return selected;
  }
  for (const pack of STYLE_PACKS.slice(1)) {
    if (!selected.includes(pack)) selected.push(pack);
    if (selected.length >= MAX_PACKS) break;
  }
  return selected;
}

function composedCount(): number {
  const raw = process.env.STUDIO_COMPOSED_VARIANTS ?? "2";
  const n = parseInt(raw, 10);
  return Number.isFinite(n) ? Math.max(0, Math.min(n, 3)) : 2;
}

async function composedTemplates(
  projectId: string,
  profile: Record<string, unknown>,
  assets: Record<string, unknown>[],
  genre: string,
): Promise<Record<string, unknown>[]> {
  const out: Record<string, unknown>[] = [];
  try {
    const composed = await compositionService.composeProjectVariants(
      projectId,
      businessFacts(profile),
      genre,
      composedCount(),
    );
    for (let idx = 0; idx < composed.length; idx++) {
      const comp = composed[idx]!;
      try {
        const html = personalizeHtml(comp.html, profile, assets, genre);
        out.push({
          templateId: `composed-${idx + 1}`,
          label: `Composed ${idx + 1} · ${genre.charAt(0).toUpperCase()}${genre.slice(1)}`,
          style: genre,
          sourceTemplateRef: comp.componentIds.join(","),
          s3HtmlKey: persistKey(projectId, `composed-${idx + 1}`),
          htmlContent: html,
        });
      } catch (exc) {
        if (exc instanceof PersonalizationError) continue;
        throw exc;
      }
    }
  } catch {
    /* composed variants optional */
  }
  return out;
}

export async function generateForProject(
  projectId: string,
  force = false,
): Promise<Record<string, unknown>[]> {
  const project = await projectsRepo.get(projectId);
  if (!project) throw new Error("project not found");
  const existing = await templatesRepo.listForProject(projectId);
  if (existing.length && !force) return existing;

  const profile = await profilesRepo.get(projectId);
  const assets = await assetsRepo.listForProject(projectId);
  const genre = pickTemplate(genreHint(project, profile));
  const templates: Record<string, unknown>[] = [];

  const composed = await composedTemplates(projectId, profile, assets, genre);
  if (composed.length) {
    for (let i = 0; i < composed.length; i++) {
      templates.push({ ...composed[i]!, sourceKind: "composed", galleryIndex: i });
    }
    const baseHtml = String(composed[0]!.htmlContent);
    for (let j = 0; j < selectedStylePacks(profile).length; j++) {
      const pack = selectedStylePacks(profile)[j]!;
      const packed = applyPack(baseHtml, pack);
      if (PLACEHOLDER_RE.test(packed)) throw new PersonalizationError("unresolved placeholders in style pack");
      const templateId = `${genre}-${pack.id}`;
      templates.push({
        templateId,
        label: `${pack.label} · ${genre.charAt(0).toUpperCase()}${genre.slice(1)}`,
        style: genre,
        sourceTemplateRef: composed[0]!.sourceTemplateRef,
        s3HtmlKey: persistKey(projectId, templateId),
        htmlContent: packed,
        sourceKind: "pack",
        galleryIndex: composed.length + j,
      });
    }
  } else {
    const designFiles = templateFiles(genre);
    if (designFiles.length) {
      const fname = designFiles[0]!;
      const stem = fname.slice(0, -".html".length);
      const raw = fs.readFileSync(path.join(TEMPLATES_DIR, fname), "utf8");
      const html = personalizeHtml(raw, profile, assets, genre);
      templates.push({
        templateId: stem,
        label: designLabel(raw, stem, genre),
        style: genre,
        sourceTemplateRef: fname,
        s3HtmlKey: persistKey(projectId, stem),
        htmlContent: html,
        sourceKind: "fallback",
        galleryIndex: 0,
      });
    }
  }

  templates.sort(
    (a, b) =>
      ((a.galleryIndex as number | undefined) ?? 999) - ((b.galleryIndex as number | undefined) ?? 999) ||
      String(a.templateId).localeCompare(String(b.templateId)),
  );

  const saved = await templatesRepo.replaceForProject(projectId, templates);
  await projectsRepo.updateFields(projectId, { status: "ready", templateCount: saved.length });
  return saved;
}
